BC marginal price

guilpier-code opened the pull request:
This PR covers : 
- In the output static list, inserting  an output variable class associated to binding constraints.
- This PR aims at handling the year by year results, but not the synthesis results.
- N/A status in year by year results should be handled correctly : for a weekly BC, hourly results are N/A 

TODO
- [x] Make sure that new files are created & written
- [x] Handle constraints of type "< and >". Add direction as 2nd line ?
- [x] Synthesis

flomnes committed:
Allocate/deallocate memory to store dual values

flomnes committed:
Save address of dual variables for binding constraints

flomnes committed:
Use alias variable

flomnes committed:
Add binding constraints constants & file names

guilpier-code committed:
bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

guilpier-code committed:
BC marginal price : make previous commit compile

guilpier-code committed:
BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

guilpier-code committed:
BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

guilpier-code committed:
BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

guilpier-code committed:
BC marginal price : computing marginal price annual statistics

guilpier-code committed:
BC marginal price : avoid too much code duplication when computing marginal price annual statistics

guilpier-code committed:
BC marginal price : set N/A status for a BC output variable depending on BC type and file level

flomnes commented on src/solver/variable/economy/bindingConstraintsMarginalCost.h:
```diff
@@ -0,0 +1,339 @@
+        }
+    }
+
+    void hourBegin(unsigned int hourInTheYear)
+    {
+        if (associatedBC_->type == Data::BindingConstraint::typeHourly)
+        {
+            pValuesForTheCurrentYear[yearMemorySpace_][hourInTheYear]
+              -= state_->problemeHebdo->ResultatsContraintesCouplantes[bindConstraintGlobalNumber_]
+                   .variablesDuales[state_->hourInTheWeek];
+        }
+
+        // Next variable
+        NextType::hourBegin(hourInTheYear);
+    }
```
Note : `hourEnd` takes the state as a parameter. There's no need to save a pointer to it if you use `hourEnd` instead of `hourBegin`.

guilpier-code replied:
Done

guilpier-code committed:
BC marginal price : printing year by year results

guilpier-code commented on src/solver/variable/commons/join.h:
```diff
@@ -210,42 +221,50 @@ class Join : public Variable::IVariable<Join<LeftT, RightT>, Yuni::Default, VCar
                         unsigned int nbYearsForCurrentSummary)
     {
         LeftType::computeSummary(numSpaceToYear, nbYearsForCurrentSummary);
+        BindConstType::computeSummary(numSpaceToYear, nbYearsForCurrentSummary);
```
**Caution** : there is an error here : function **BindConstType::computeSummary** is called twice.
This error is fixed in a further PR (which is : #812 )

flomnes commented on src/solver/cmake/variable.cmake:
```diff
@@ -16,6 +16,8 @@ set(SRC_VARIABLE
 		variable/area.inc.hxx
 		variable/setofareas.h
 		variable/setofareas.hxx
+		variable/bindConstraints.h
```
bindingConstraint

flomnes commented on src/solver/cmake/variable.cmake:
```diff
@@ -142,6 +146,7 @@ set(SRC_VARIABLE_ECONOMY
 		variable/economy/lolp.h
 		variable/economy/avail-dispatchable-generation.h
 		variable/economy/dispatchable-generation-margin.h
+		variable/economy/bindingConstraintsMarginalCost.h
```
There's a dedicated directory for link-supported variables. Add one for BCs ?

guilpier-code replied:
Done

flomnes commented on src/solver/variable/bindConstraints.hxx:
```diff
@@ -0,0 +1,616 @@
+    // This actually a vector of pairs where :
+    //  - first element of a pair is a pointer to a inequality binding constraint
+    //  - second element of a pair is the number of the previous BC in the runtime list of BCs.
+    std::vector<std::pair<Data::BindingConstraintRTI*, uint>> InequalityBCs
```
If we want to filter constraints (only < or >), a vector of indices probably does the job here. Though I'm not sure we want to filter constraints. TBD with users & POs

guilpier-code replied:
"a vector of indices probably does the job here" ==> done

flomnes replied:
Where ?

guilpier-code replied:
Finally done

flomnes reviewed (changes requested):
A lot of commented-out code, please do some cleaning

guilpier-code commented:
For code cleaning you ask, see PR #824

guilpier-code committed:
BC marginal price : changes after review

guilpier-code committed:
BC marginal price : correction on previous commit

guilpier-code committed:
BC marginal price : fixes so that compilation passes

guilpier-code committed:
BC marginal price : correction due to sonarQube review

guilpier-code committed:
BC marginal price : making compilation pass

flomnes commented on src/solver/variable/economy/bindingConstraints/bindingConstraintsMarginalCost.h:
```diff
@@ -28,7 +28,7 @@
 
 #include <math.h>
```
```suggestion
#include <cmath>
```

guilpier-code replied:
Done

flomnes commented on src/solver/variable/economy/bindingConstraints/bindingConstraintsMarginalCost.h:
```diff
@@ -0,0 +1,321 @@
+    unsigned int pNbYearsParallel;
+    Data::BindingConstraintRTI* associatedBC_ = nullptr;
+    uint yearMemorySpace_ = 0;
+    uint bindConstraintGlobalNumber_ = -1;
```
This will lead to invalid reads in arrays. Make sure to check that `bindConstraintGlobalNumber_>=0` before reading e.g `study.runtime->bindingConstraint[bindConstraintGlobalNumber_]`

guilpier-code replied:
Handled

flomnes commented on src/solver/variable/storage/intermediate.cpp:
```diff
@@ -241,6 +241,54 @@ void IntermediateValues::computeAVGstatisticsForCurrentYear()
+
+void IntermediateValues::computeAnnualAveragesFromWeeklyValues()
+{
+    uint i;
+    uint j;
```
```suggestion
    uint i, j;
```

guilpier-code replied:
Done

guilpier-code committed:
BC marginal price : enhancement due to sonar review

flomnes commented on src/solver/variable/economy/bindingConstraints/bindingConstraintsMarginalCost.h:
```diff
@@ -135,13 +135,12 @@ class BindingConstMarginCost
     ~BindingConstMarginCost()
     {
-        delete[] pValuesForTheCurrentYear;
+        if (!pValuesForTheCurrentYear)
```
```suggestion
        if (pValuesForTheCurrentYear)
```

guilpier-code replied:
Done

guilpier-code committed:
BC marginal price : changes due to human review

guilpier-code committed:
BC marginal price : making previous commit compile on Linux

guilpier-code committed:
BC marginal price : another try to make compilation pass on Linux

guilpier-code committed:
BC marginal price : small (but important) correction + new try to make compilation pass on Linux

guilpier-code committed:
BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

guilpier-code committed:
BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

flomnes committed:
Merge remote-tracking branch 'github/develop' into HEAD
